SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc

```diff
@@ -11,27 +11,300 @@
+{
+  HGCSimHitData chargeColl, toa;
+  // this represents a cell with no signal charge
+  HGCCellInfo zeroData;
+  zeroData.hit_info[0].fill(0.f); //accumulated energy
+  zeroData.hit_info[1].fill(0.f); //time-of-flight
+
+  for( const auto& id : validIds ) {
+
+    chargeColl.fill(0.f);
+    toa.fill(0.f);
+    HGCSimHitDataAccumulator::iterator it = simData.find(id);
+    HGCCellInfo& cell = ( simData.end() == it ? zeroData : it->second );
+    addCellMetadata(cell,theGeom,id);
+
+    for(size_t i=0; i<cell.hit_info[0].size(); ++i)
+    {
+      //convert total energy keV->MIP, since converted to keV in accumulator
+      const float totalIniMIPs( cell.hit_info[0][i]*keV2MIP_ );
+
+      //store
+      chargeColl[i] = totalIniMIPs;
+    }
+
+    //init a new data frame and run shaper
+    HGCalDataFrame newDataFrame( id );
+    this->myFEelectronics_->runShaper( newDataFrame, chargeColl, toa, 1, engine );
+
+    //prepare the output
+    this->updateOutput(digiColl,newDataFrame);
+  }
+}
+
+void HGCHEbackDigitizer::runRealisticDigitizer(std::unique_ptr<HGCalDigiCollection> &digiColl,HGCSimHitDataAccumulator &simData,
+  const CaloSubdetectorGeometry* theGeom, const std::unordered_set<DetId>& validIds,
+  CLHEP::HepRandomEngine* engine)
+  {
+    //switch to true if you want to print some details
+    constexpr bool debug(false);
+
+    HGCSimHitData chargeColl, toa;
+    // this represents a cell with no signal charge
+    HGCCellInfo zeroData;
+    zeroData.hit_info[0].fill(0.f); //accumulated energy
+    zeroData.hit_info[1].fill(0.f); //time-of-flight
+
+    // needed to compute the radiation and geometry scale factors
+    scal_.setGeometry(theGeom);
+
+    for( const auto& id : validIds ) {
+
+      chargeColl.fill(0.f);
+      toa.fill(0.f);
+      HGCSimHitDataAccumulator::iterator it = simData.find(id);
+      HGCCellInfo& cell = ( simData.end() == it ? zeroData : it->second );
+      addCellMetadata(cell,theGeom,id);
+
+      for(size_t i=0; i<cell.hit_info[0].size(); ++i)
+      {
+        //convert total energy keV->MIP, since converted to keV in accumulator
+        float totalIniMIPs( cell.hit_info[0][i]*keV2MIP_ );
+
+        //take into account the different size of the tiles
+        float scaledPePerMip = nPEperMIP_;
+        if(scaleByArea_)
+          scaledPePerMip *= scal_.scaleByArea(id);
+
+        //take into account the darkening of the scintillator and SiPM dark current
+        float tunedNoise = nPEperMIP_ * noise_MIP_; //flat noise case
+        if(scaleByDose_)
+        {
+          auto dosePair = scal_.scaleByDose(id);
+          scaledPePerMip *= dosePair.first;
+          tunedNoise = dosePair.second;
+        }
+
+        //generate the number of photo-electrons from the energy deposit
+        const uint32_t npeS = std::floor(CLHEP::RandPoissonQ::shoot(engine, totalIniMIPs * scaledPePerMip) + 0.5);
+
+        //generate the noise associated to the dark current
+        float meanN = std::pow(tunedNoise, 2);
```

Comment:
similar comment about repeated `pow` (the default `tunedNoise`^2 could be defined outside the loop)